### PR TITLE
Philipp lig 1511 implement smog

### PR DESCRIPTION
**Pull request by philippmwirth**
# DRAFT: Add SMoG

**This is outdated, see the comments below for the current state of the PR**

Draft implementation of SMoG: https://arxiv.org/pdf/2207.06167.pdf

Currently, the CIFAR10 benchmark reaches approximately `65%` accuracy after 200 epochs (depending on the exact set of hyperparameters). 

There are some questions and problems I'm encountering:

- The loss behaves strangely. After resetting the cluster centers, it grows quickly and then starts going down again.
- In the paper, they clearly state that they use a _prediction head_ on top of the momentum network:
image
    In my opinion this doesn't make sense because there's no gradient going through the momentum network. I just left it away in this draft.
- There's a possible typo in the paper:
image
    I used `argmax` instead.
- Finally, they mention in the paper that they're using an asymmetric version of the BYOL augmentations where one of the augmentations uses stronger solarization and the other stronger Gaussian blur. However, there's no further information about this in the paper so I simply used SimCLR augmentations.

Note: The repeated KMeans clustering slows down training quite a bit.

## TODO in follow-up issues

- CIFAR10 benchmark
- Imagenet100 benchmark

### CODE REVIEW

**Comment by philippmwirth** [comment]
> Sorry for the typos.
> 
>     1. The prediction head is added on f_\theta.
> 
>     2. argmin -> argmax in Eq.4.
> 
>     3. SimCLR's augmentation is OK for SMoG.

Thanks a lot for the help @BoPang1996 🙂 I will try with your suggestions and post an update.

File: lightly/models/modules/smog.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        Args:
+            Tensor of shape bsz x dim.
+
+        Returns:
+            LongTensor of shape bsz indicating group assignments.
+        
+        """
+        x = torch.nn.functional.normalize(x)
+        group_features = torch.nn.functional.normalize(self.group_features)
+        logits = torch.mm(x, group_features.t())
```
Comment (guarin) [inline lightly/models/modules/smog.py:84]: I would put everything up to and including the logits calculation into the `forward` function of the module. And rewrite `assign_groups` to:
```
return torch.argmax(self.forward(x), dim=1)
```

This allows the forward function to be reused for this line in the benchmark module:
```
logits = torch.mm(torch.nn.functional.normalize(x0_predicted), group_features.t()) / 0.5
```

File: docs/source/getting_started/benchmarks/cifar10_benchmark.py [hunk 2]
```diff
@@ -729,17 +729,113 @@ def configure_optimizers(self):
...
+        utils.deactivate_requires_grad(self.projection_head_momentum)
+
+    def training_step(self, batch, batch_idx):
+
+        if batch_idx == 0:
+            # the original paper reset group features and weights every 300
+            # iterations, here we do it once an epoch (TODO)
+            self._reset_group_features()
+            self._reset_momentum_weights()
+            pass
```
Comment (guarin) [inline docs/source/getting_started/benchmarks/cifar10_benchmark.py:787]: Remove pass

File: docs/source/getting_started/benchmarks/cifar10_benchmark.py [hunk 3]
```diff
@@ -729,17 +729,113 @@ def configure_optimizers(self):
...
+        else:
+            # update momentum
+            utils.update_momentum(self.backbone, self.backbone_momentum, 0.99)
+            utils.update_momentum(self.projection_head, self.projection_head_momentum, 0.99)
+
+        (x0, x1), _, _ = batch
```
Comment (guarin) [inline docs/source/getting_started/benchmarks/cifar10_benchmark.py:793]: I think you forgot to use the `SMoGCollateFunction` :)

File: docs/source/getting_started/benchmarks/cifar10_benchmark.py [hunk 4]
```diff
@@ -729,17 +729,113 @@ def configure_optimizers(self):
...
+
+        # update group features and get group assignments
+        assignments = self.smog.assign_groups(x1_encoded)
+        group_features = self.smog.update_groups(x0_encoded)
+
+        logits = torch.mm(torch.nn.functional.normalize(x0_predicted), group_features.t()) / 0.5
+        loss = self.criterion(logits, assignments)
+
+        # use memory bank to periodically reset the group features with k-means
+        self.memory_bank(x0_encoded.detach(), update=True)
```
Comment (guarin) [inline docs/source/getting_started/benchmarks/cifar10_benchmark.py:809]: Detach is already handled by memory bank

File: lightly/models/modules/smog.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    
+    """
+
+    def __init__(
+        self, n_groups: int, dim: int, beta: float, device: Optional[str] = None
+    ):
+        super(SMoG, self).__init__()
+        self.n_groups = n_groups
+        self.dim = dim
+        self.group_features = torch.rand(n_groups, dim).to(device)
```
Comment (guarin) [inline lightly/models/modules/smog.py:28]: I would maybe make `self.group_features` a parameter of the module. Then you can also easily save it / move it to other devices.

File: docs/source/getting_started/benchmarks/cifar10_benchmark.py (hunk 3 above)
Comment (philippmwirth) [reply docs/source/getting_started/benchmarks/cifar10_benchmark.py:793]: This is on purpose to keep it comparable with the other strategies. But I'll try it and report back.

File: docs/source/getting_started/benchmarks/cifar10_benchmark.py [hunk 6]
```diff
@@ -729,17 +729,113 @@ def configure_optimizers(self):
...
+        x0_encoded = self.projection_head(x0_features)
+        x0_predicted = self.prediction_head(x0_encoded)
+        x1_features = self.backbone_momentum(x1).flatten(start_dim=1)
+        x1_encoded = self.projection_head_momentum(x1_features)
+
+        # update group features and get group assignments
+        assignments = self.smog.assign_groups(x1_encoded)
+        group_features = self.smog.update_groups(x0_encoded)
+
+        logits = torch.mm(torch.nn.functional.normalize(x0_predicted), group_features.t()) / 0.5
```
[the remark this answers is not in the document]
Comment (philippmwirth) [reply docs/source/getting_started/benchmarks/cifar10_benchmark.py:805]: It's the temperature parameter.

[the remark this answers is not in the document]
Comment (guarin) [reply docs/source/getting_started/benchmarks/cifar10_benchmark.py:805]: AFAIK temperature is 0.1 in the paper.

[the remark this answers is not in the document]
Comment (philippmwirth) [reply docs/source/getting_started/benchmarks/cifar10_benchmark.py:805]: yes but they're not running the experiments on cifar10 with a resnet18. 0.5 and 0.1 both work fine

File: lightly/models/modules/smog.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+        Args:
+            Tensor of shape bsz x dim.
+
+        Returns:
+            LongTensor of shape bsz indicating group assignments.
+        
+        """
+        x = torch.nn.functional.normalize(x)
+        group_features = torch.nn.functional.normalize(self.group_features)
```
[the remark this answers is not in the document]
Comment (philippmwirth) [reply lightly/models/modules/smog.py:83]: I wouldn't make this assumption when calculating the cosine similarity.

File: lightly/models/modules/smog.py (hunk 5 above)
Comment (philippmwirth) [reply lightly/models/modules/smog.py:28]: This doesn't work because it requires a gradient but only for the "added" part.

Comment (guarin) [reply lightly/models/modules/smog.py:28]: I fear I don't understand. Parameters can have gradients. And you can always disable them temporarily with `torch.no_grad`.

Comment (philippmwirth) [reply lightly/models/modules/smog.py:28]: I'm trying out some stuff. The problem is that you do a momentum update and you only want the gradient only for the "update" part.

File: docs/source/getting_started/benchmarks/cifar10_benchmark.py [hunk 8]
```diff
@@ -729,17 +740,118 @@ def configure_optimizers(self):
...
+        utils.deactivate_requires_grad(self.backbone_momentum)
+        utils.deactivate_requires_grad(self.projection_head_momentum)
+
+        # smog
+        self.n_groups = 300 # 2% malus vs optimal setting of 3000 groups
+        self.memory_bank = lightly.loss.memory_bank.MemoryBankModule(size=153600)
+        # create our loss
+        group_features = torch.nn.functional.normalize(
+            torch.rand(self.n_groups, 128), dim=1
+        ).cuda()
```
Comment (guarin) [inline docs/source/getting_started/benchmarks/cifar10_benchmark.py:773]: `.to(self.device)`

File: docs/source/getting_started/benchmarks/cifar10_benchmark.py [hunk 9]
```diff
@@ -729,17 +740,118 @@ def configure_optimizers(self):
...
+
+    def _reset_group_features(self):
+        # see Table 7b)
+        features = self.memory_bank.bank
+        if features is not None:
+            features = features.t().cpu().numpy()
+            kmeans = KMeans(self.n_groups).fit(features)
+            new_features = torch.from_numpy(kmeans.cluster_centers_).float()
+            new_features = torch.nn.functional.normalize(new_features, dim=1)
+            self.smog.group_features = new_features.cuda()
```
Comment (guarin) [inline docs/source/getting_started/benchmarks/cifar10_benchmark.py:785]: `.to(self.device)`

File: docs/source/getting_started/benchmarks/cifar10_benchmark.py [hunk 10]
```diff
@@ -729,17 +740,118 @@ def configure_optimizers(self):
...
+        # see Table 7b)
+        self.backbone_momentum = copy.deepcopy(self.backbone)
+        self.projection_head_momentum = copy.deepcopy(self.projection_head)
+        utils.deactivate_requires_grad(self.backbone_momentum)
+        utils.deactivate_requires_grad(self.projection_head_momentum)
+
+    def training_step(self, batch, batch_idx):
+
+        if self.global_step > 0 and self.global_step % 300 == 0:
+            # reset group features and weights every 300
```
Comment (guarin) [inline docs/source/getting_started/benchmarks/cifar10_benchmark.py:797]: --> every 300 iterations

**Review by guarin — APPROVED** [review]
Changes look great! I left some conversations open for better visibility, feel free to close them when merging :)

File: lightly/models/modules/heads.py [hunk 11]
```diff
@@ -223,6 +223,99 @@ def __init__(self,
...
+
+    [0]: SMoG, 2022, https://arxiv.org/pdf/2207.06167.pdf
+    
+    """
+    def __init__(self,
+                 input_dim: int = 128,
+                 hidden_dim: int = 2048,
+                 output_dim: int = 128):
+        super(SMoGPredictionHead, self).__init__([
+            (input_dim, hidden_dim, nn.BatchNorm1d(2048), nn.ReLU()),
```
Comment (MalteEbner) [inline lightly/models/modules/heads.py:313]: @guarin 
Why is this 2048 and not hidden_dim?

Comment (guarin) [reply lightly/models/modules/heads.py:313]: Looks like an error :) I agree that it should be hidden dim.

File: docs/source/getting_started/benchmarks/imagenette_benchmark.py [hunk 12]
```diff
@@ -21,6 +21,7 @@
 | DCLW (*)      |        256 |    200 |              0.755 |   53.7 Min |      4.3 GByte |
 | DINO (Res18)  |        256 |    200 |              0.736 |   86.5 Min |      4.1 GByte |
 | MSN (ViT-S)   |        256 |    200 |              0.741 |   92.7 Min |     16.3 GByte |
+| SMoG          |        128 |    200 |              0.698 |  220.9 Min |     14.3 GByte |
```
Comment (guarin) [inline docs/source/getting_started/benchmarks/imagenette_benchmark.py:24]: Could you add the result between SimSiam and SwaV to keep alphabetical ordering?

Comment (guarin) [reply docs/source/getting_started/benchmarks/imagenette_benchmark.py:24]: Also, is it not possible to train with batch size 256?

**Review by guarin — APPROVED** [review]

File: docs/source/getting_started/benchmarks/imagenette_benchmark.py (hunk 12 above)
Comment (MalteEbner) [reply docs/source/getting_started/benchmarks/imagenette_benchmark.py:24]: Probably a batch size of 256 would need too much memory.

File: lightly/models/modules/heads.py (hunk 11 above)
Comment (MalteEbner) [reply lightly/models/modules/heads.py:313]: I changed it.
